### ISSUE TITLE
XGBoost has trouble modeling multiplication/division

### ISSUE THREAD

**Issue by ledmaster** [dmlc/xgboost#4069]
Hi

I am using Python 3.6 and XGBoost version 0.81. When I try a simple experiment, creating a matrix X of numbers between 1 and -1, and then Y = X1 * X2 or Y = X1 / X2, xgboost can't learn and predicts a constant number.

![multiplication_noise0](https://user-images.githubusercontent.com/9610024/51441494-7640da80-1cb9-11e9-9a51-a9638b27850a.jpg)

Now, if I add gaussian noise, it can model the function:

![multiplication_noise05](https://user-images.githubusercontent.com/9610024/51441500-80fb6f80-1cb9-11e9-9a18-b856034b17f0.jpg)

I tried changing the range, tuning hyperparameters, base_score, using the native xgb.train vs XGBoostRegressor, but couldn't make it learn.

Is this a known issue? Do you know why it happens?

Thanks

**Comment by trivialfis** [comment]
@ledmaster I tried to generate the following dataset, is it the right one?
``` python
x = np.random.rand(64, 2)
x = x * 2 - 1.0
y_true = x[:, 0] * x[:, 1]
```

Following above script:
``` python
dtrain = xgb.DMatrix(x, label=y_true)

params = {
    'tree_method': 'gpu_hist'
}

bst = xgb.train(params, dtrain, evals=[(dtrain, "train")], num_boost_round=10)

y_pred = bst.predict(dtrain)
# Z = pred_y

X = x[:, 0]
Y = x[:, 1]

fig = plt.figure(figsize=plt.figaspect(2.))
ax = fig.add_subplot(2, 1, 1, projection='3d')
ax.plot_trisurf(X, Y, y_pred, cmap='viridis')

ax = fig.add_subplot(2, 1, 2, projection='3d')
ax.plot_trisurf(X, Y, y_true, cmap='viridis')

plt.show()
```

I got:

![figure_1](https://user-images.githubusercontent.com/16746409/51477127-6b25a180-1dc2-11e9-8f57-c6c496aa573b.png)


Seems pretty reasonable.  Did I generate the wrong dataset?

**Comment by trivialfis** [comment]
I lose.  Adding small normal noise (`np.random.randn()`) to `label` everything works, but without such noise xgboost just jump into some sort of local minimum.
``` python
        noise = np.random.randn(size)
        noise = noise / 1000
        y = y + noise  # somehow this help XGBoost leave the local minimum
        dtrain = xgb.DMatrix(X, label=y)
```
I'm not sure if this is a bug.  XGBoost relies on greedy algorithm after all.  Would love to hear some other opinions. @RAMitchell

**Comment by khotilov** [comment]
I agree with @trivialfis - it's not a bug. But it's a nice example of data with "perfect symmetry" with unstable balance. With such perfect dataset, when the algorithm is looking for a split, say in variable X1, the sums of residuals at each x1 location of it WRT X2 are always zero, thus it cannot find any split and is only able to approximate the total average. Any random disturbance, e.g., some noise or `subsample=0.8` would help to kick it off the equilibrium and to start learning.

**Comment by trivialfis** [comment]
@khotilov That's a very interesting example.  I wouldn't come up with it myself.  Maybe we can document it in a tutorial?

**Comment by hcho3** [comment]
@ledmaster Thanks for the article: http://mariofilho.com/can-gradient-boosting-learn-simple-arithmetic/

**Comment by ledmaster** [comment]
Thanks @khotilov and @trivialfis for the answers and investigation.

As @hcho3 cited above, I wrote an article about GBMs and arithmetic operations, so this is why I ended up finding the issue. I added your answer there. Feel free to link to it.

**Comment by RAMitchell** [comment]
Decision trees can only look at one feature at a time. In your example if I take any 1d feature range such as 0.25 < f0 < 0.5 and average all the samples in that slice I suspect you will always get exactly 0 due to the symmetry of your problem. So xgboost cant find anything when it looks from the perspective of a single feature.
